nowcasting_dataset/dataset/xr_utils.py

```diff
@@ -11,48 +11,64 @@
+# Maybe we should move this function to fake.py?
+def join_list_data_array_to_batch_dataset(data_arrays: List[xr.DataArray]) -> xr.Dataset:
+    """Join a list of xr.DataArrays into an xr.Dataset by concatenating on the example dim."""
+    data_arrays = [convert_data_array_to_dataset(data_arrays[i]) for i in range(len(data_arrays))]
 
-    return join_dataset_to_batch_dataset(image_data_arrays)
+    return join_list_dataset_to_batch_dataset(data_arrays)
 
 
-def join_dataset_to_batch_dataset(image_data_arrays: List[xr.Dataset]) -> xr.Dataset:
-    """ Join a list of data arrays to a dataset byt expanding dims """
-    image_data_arrays = [
-        image_data_arrays[i].expand_dims(dim="example").assign_coords(example=("example", [i]))
-        for i in range(len(image_data_arrays))
-    ]
+def join_list_dataset_to_batch_dataset(datasets: list[xr.Dataset]) -> xr.Dataset:
+    """ Join a list of data sets to a dataset by expanding dims """
 
-    return xr.concat(image_data_arrays, dim="example")
+    new_datasets = []
+    for i, dataset in enumerate(datasets):
+        new_dataset = dataset.expand_dims(dim="example").assign_coords(example=("example", [i]))
+        new_datasets.append(new_dataset)
 
+    return xr.concat(new_datasets, dim="example")
 
+
+# TODO: Issue #318: Maybe remove this function and, in calling code, do data_array.to_dataset()
+# followed by make_dim_index, to make it more explicit what's happening?  At the moment,
+# in the calling code, it's not clear that the coordinates are being changed.
 def convert_data_array_to_dataset(data_xarray):
     """ Convert data array to dataset. Reindex dim so that it can be merged with batch"""
     data = xr.Dataset({"data": data_xarray})
-
-    return make_dim_index(data_xarray_dataset=data)
+    return make_dim_index(dataset=data)
 
 
-def make_dim_index(data_xarray_dataset: xr.Dataset) -> xr.Dataset:
-    """ Reindex dataset dims so that it can be merged with batch"""
+# TODO: Issue #318: Maybe rename this function... maybe to coord_to_range()?
+# Not sure what's best right now!  :)
+def make_dim_index(dataset: xr.Dataset) -> xr.Dataset:
+    """Reindex dims so that it can be merged with batch.
 
-    dims = data_xarray_dataset.dims
+    For each dimension in dataset, change the coords to 0.. len(original_coords),
+    and append "_index" to the dimension name.
+    And save the original coordinates in `original_dim_name`.
 
-    for dim in dims:
-        coord = data_xarray_dataset[dim]
-        data_xarray_dataset[dim] = np.arange(len(coord))
-
-        data_xarray_dataset = data_xarray_dataset.rename({dim: f"{dim}_index"})
+    This is useful to align multiple examples into a single batch.
+    """
 
-        data_xarray_dataset[dim] = xr.DataArray(
-            coord, coords=data_xarray_dataset[f"{dim}_index"].coords, dims=[f"{dim}_index"]
+    original_dim_names = dataset.dims
+
+    for original_dim_name in original_dim_names:
```

Comment:
This new code should be functionally equivalent to the old code!  I just wanted to make it really explicit what's going on.